### DISCUSSION TITLE
class `Shape`  and  less reflection verification

### DISCUSSION THREAD

**Discussion by John0King** [dotnet/csharplang#6336] [Language Ideas]
In TypeScript there is `Parital<T>` and other related type for help better desgin the api.
My idea is to have `shape` as the type's signature， 
```C#
//simple shape
public shape FooShape
{
    string Name{get;set;};
    int Age{ get; set; }
}

// complex shape
public shape Partial<T> where T:class
{
    [keyof T keyName]: T.keyName;
}

public class Foo
{
    public string Name{get;set;}
    public int Age{ get;set; }
}

// using shape

public void DoSomething(object data) 
    when data: FooShape
{
     var age = data.Age;   // interal reflect  call  but do not have to verifiy the type or name
     var name = data.Name; // interal reflect  call  but do not have to verifiy the type or name
     data.Age=  age + 1;  // interal reflect  call  but do not have to verifiy the type or name

}

DoSomething(new Foo())

```

1. type do **NOT** have to implment the shape (but can) , shape will be verified  at compile time.
2. a internal less verification version of reflect should be add to corefx , so it can leverage  compile time verification, so it should be fast like emit;

**Comment by huoyaoyuan** [comment]
The type system of IL works very differently with javascript.
You can't invoke a member that's not associated with realistic interface/class.


> 2\. a internal less verification version of reflect should be add to corefx , so it can leverage  compile time verification, so it should be fast like emit;

This is not achievable under current type system. There must be massive change in CLR.
You can see more discussion of similar request in #5496

**Reply by John0King** [reply]
> You can't invoke a member that's not associated with realistic interface/class.

this invoke is base on reflection,  eg.  
```C#
data.GetType()
    .FastGetProperty(propertyName: "Name", propertyType: typeof(string),  getSetFlag: getSetFlag.GetAndSet)
    .FastSet<string>("abc")

```
